dotcom-rendering/src/lib/personaliseHighlights.ts

```diff
@@ -0,0 +1,171 @@
+ * If editorial has updated the highlights container, we reset the ordering to allow for editorial oversight.
+ * The user only stores one highlight container history in storage. If a different front is visited that has a highlights container,
+ * it will reset local storage with the new container data.
+ * */
+
+type HighlightCardHistory = {
+	card: DCRFrontCard; // TODO: store a card identifier (eg url) rather than the whole card.
+	viewCount: number;
+	wasClicked: boolean;
+};
+
+export type HighlightHistory = Array<HighlightCardHistory>;
+
+type CardEngagement = 'VIEW' | 'CLICK';
+
+export const HighlightsHistoryKey = 'gu.history.highlights';
+
+const MAX_VIEW_COUNT = 3;
+
+const isValidHighlightHistory = (
+	history: unknown,
+): history is HighlightHistory =>
+	Array.isArray(history) &&
+	history.every(
+		(highlight) =>
+			isObject(highlight) &&
+			'card' in highlight &&
+			'viewCount' in highlight &&
+			'wasClicked' in highlight &&
+			isObject(highlight.card) &&
+			typeof highlight.viewCount === 'number' &&
+			typeof highlight.wasClicked === 'boolean',
+	);
+
+// Retrieve the user's highlight card order from local storage
+export const getHighlightHistory = (): HighlightHistory | undefined => {
+	try {
+		const highlightHistory = storage.local.get(HighlightsHistoryKey);
+
+		if (!isValidHighlightHistory(highlightHistory)) {
+			throw new Error(`Invalid ${HighlightsHistoryKey} value`);
+		}
+
+		return highlightHistory;
+	} catch (e) {
+		// error parsing the string, so remove the key
+		storage.local.remove(HighlightsHistoryKey);
+		return undefined;
+	}
+};
+
+// remove highlight history from local storage
+const removeHighlightHistory = (): void => {
+	storage.local.remove(HighlightsHistoryKey);
+};
+
+// store the personalised history in local storage
+export const storeHistoryInStorage = (order: HighlightHistory): void => {
+	storage.local.set(HighlightsHistoryKey, order);
+};
+
+// Maps DCR front cards to history records, initialising view and click tracking
+export const convertCardsToHistory = (
+	cards: Array<DCRFrontCard>,
+): HighlightHistory => {
+	return cards.map((card) => ({
+		card,
+		viewCount: 0,
+		wasClicked: false,
+	}));
+};
+
+// Reset highlight history in local storage
+export const resetStoredHighlights = (cards: DCRFrontCard[]): void => {
+	removeHighlightHistory();
+	const highlights = convertCardsToHistory(cards);
+	storeHistoryInStorage(highlights);
+};
+
+// Maps history records to DCR front cards for faster rendering
+export const getCardsFromHistory = (
+	history: HighlightHistory,
+): Array<DCRFrontCard> => {
+	return history.map((highlight) => {
+		return highlight.card;
+	});
+};
+
+export const getHighlightCards = (): Array<DCRFrontCard> => {
+	const history = getHighlightHistory() ?? [];
+	return getCardsFromHistory(history);
+};
+
+// Track when a user has clicked on a highlight card
```

Comment:
Comments for functions are better in the `/** I'm a comment */` format rather than `// I'm a comment`, as you get to see the comment whenever you hover over the function where it is called in VS Code